daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -1099,31 +1098,82 @@ private[lf] object Speedy {
             }
           case TTyCon(tyCon) =>
             value match {
-              case V.ValueRecord(_, fields) =>
-                val lookupResult =
-                  assertRight(compiledPackages.pkgInterface.lookupDataRecord(tyCon))
+              case V.ValueRecord(_, sourceElements) => { // This _ is the source typecon, which we ignore.
+                val lookupResult = assertRight(
+                  compiledPackages.pkgInterface.lookupDataRecord(tyCon)
+                )
+                val targetFieldsAndTypes: ImmArray[(Name, Type)] = lookupResult.dataRecord.fields
                 lazy val subst = lookupResult.subst(argTypes)
-                // TODO: https://github.com/digital-asset/daml/issues/17082
-                // - disallow dropping extra fields which have value other than None.
-                val actualValues = (lookupResult.dataRecord.fields.iterator zip fields.iterator)
-                  .map { case ((_, fieldType), (_, fieldValue)) =>
-                    go(AstUtil.substitute(fieldType, subst), fieldValue)
+
+                // This code implements the compatibility transformation used for up/down-grading
+                // And handles the cases:
+                // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
+                //
+                // When numS == numT, we wont hit the code marked either as UPGRADE or DOWNGRADE,
+                // although it is still possible that the source and target types are different,
+                // but since we don't consult the source type (may be unavailable), we wont know.
+
+                val numS: Int = sourceElements.length
+                val numT: Int = targetFieldsAndTypes.length
+
+                // traverse the sourceElements, "get"ing the corresponding target type
+                // when there is no corresponding type, we must be downgrading, and so we insist the value is None
+                val values0: List[SValue] =
+                  sourceElements.toSeq.zipWithIndex.flatMap { case ((optName, v), i) =>
+                    targetFieldsAndTypes.get(i) match {
+                      case Some(x) => {
+                        val (targetField, targetFieldType): (Name, Type) = x
+                        optName match {
+                          case None => ()
+                          case Some(sourceField) =>
+                            // value is not normalized; check field names match
```

Comment:
This comment is a bit off. All value that pass through `imporValue` are actually normalized. 
Normalization of old version actually have field name.